src/world/GameWorldGame.cpp

```diff
@@ -1315,6 +1315,54 @@ void GameWorldGame::ConvertMineResourceTypes(Resource::Type from, Resource::Type
     }
 }
 
+/**
+* Fills water depending on terrain and Addon setting
+*/
+void GameWorldGame::PlaceAndFixWater()
+{
+    bool waterEverywhere = GetGGS().getSelection(AddonId::EXHAUSTIBLE_WATER) == 1;
+
+    RTTR_FOREACH_PT(MapPoint, GetSize())
+    {
+        Resource curNodeResource = GetNode(pt).resources;
+
+        if (curNodeResource.getType() != Resource::Nothing && curNodeResource.getType() != Resource::Water) {
+            // do not override maps resource.
+            continue;
+        }
+
+        int amount = 0;
+
+        // only set water if no desert, water, mountain or lava
+        if (!(World::IsOfTerrain(pt, TT_DESERT) || World::IsOfTerrain(pt, TerrainData::IsWater)
+            || World::IsOfTerrain(pt, TerrainData::IsMountain) || World::IsOfTerrain(pt, TerrainData::IsLava)))
+        {
+            
+            if (waterEverywhere)
+                amount = 7;
+            // do not touch tile if waterEverywhere is disabled and no resource was stored by maploader
+            else if (curNodeResource.getType() == Resource::Nothing) 
+                amount = 0;
+            else if (World::IsOfTerrain(pt, TT_SAVANNAH)) // reduce water on stepppe or savannah tiles.
+                amount = 4;
+            else if (World::IsOfTerrain(pt, TT_STEPPE))
+                amount = 2;
+            else
+                amount = 7;
+        }
+    
+        if (amount != 0)
+        {
+            curNodeResource.setType(Resource::Water);
```

Comment:
`curNodeResource = Resource(Resource::Water, amount)`

Reply:
Yeah true, while coding i did not realize that Resource is a struct and i was somehow confused that there is no constructor....

Reply:
Hm? There is a Ctor

Reply:
you may see that i am not programming mainly in c++. i meant there is no new operator... :-)